packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -459,6 +460,27 @@ LightningElement.prototype = {
         return getBoundingClientRect(elm);
     },
 
+    attachInternals(): ElementInternals {
+        const vm = getAssociatedVM(this);
+        const {
+            elm,
+            renderer: { attachInternals },
+        } = vm;
+
+        if (typeof ElementInternals === 'undefined') {
+            // Browsers that don't support attachInternals will need to be polyfilled before LWC is loaded.
+            throw new Error('attachInternals API is not supported in this browser environment.');
```

Comment:
I'd like to note that Jest/JSDOM will throw this error due to https://github.com/jsdom/jsdom/pull/3561 not being merged yet. That's fine; developers will have to support old Safari anyway.